File: beachechoes/modules/beachechoes-ar/ios/AprilTagDetector.swift
import ARKit

/// Wraps AprilTagBridge and adds multi-frame pose smoothing / stability gating.
class AprilTagDetector {

  /// Minimum consecutive frames a tag must be seen before it is considered stable.
  private let requiredStableFrames = 5
  /// Exponential moving average factor for pose smoothing (0–1, higher = less smoothing).
  private let smoothingAlpha: Double = 0.3
  /// Default tag size in meters (6 inches ≈ 0.1524 m).
  private let tagSizeMeters: Double = 0.1524

  private let bridge = AprilTagBridge()

  /// Per-tag tracking state for stability.
  private var tagHistory: [Int: TagTrack] = [:]

  struct TagTrack {
    var consecutiveFrames: Int = 0
    var smoothedPose: [Double]  // 16-element column-major 4x4
    var lastCenter: (x: Double, y: Double)
  }

  struct StableDetection {
    let tagId: Int
    let poseMatrix: [Double]  // 16-element column-major 4x4
    let centerX: Double
    let centerY: Double
    let hamming: Int
    let decisionMargin: Float
    let consecutiveFrames: Int
    let isNewlyStable: Bool
  }

  /// Run detection on an ARFrame and return stable detections.
  func detect(frame: ARFrame) -> [StableDetection] {
    let pixelBuffer = frame.capturedImage
    let intrinsics = frame.camera.intrinsics

    let fx = Double(intrinsics[0][0])
    let fy = Double(intrinsics[1][1])
    let cx = Double(intrinsics[2][0])
    let cy = Double(intrinsics[2][1])

    let rawResults = bridge.detectTags(
      pixelBuffer,
      fx: fx, fy: fy, cx: cx, cy: cy,
      tagSize: tagSizeMeters
    )

    var currentTagIds = Set<Int>()
    var stableDetections: [StableDetection] = []

    for result in rawResults {
      let tagId = Int(result.tagId)
      currentTagIds.insert(tagId)

      let pose = result.poseMatrix.map { $0.doubleValue }

      if var track = tagHistory[tagId] {
        track.consecutiveFrames += 1
        track.smoothedPose = smoothPose(previous: track.smoothedPose, current: pose)
        track.lastCenter = (x: result.centerX, y: result.centerY)
        tagHistory[tagId] = track

        let isNewlyStable = track.consecutiveFrames == requiredStableFrames

        if track.consecutiveFrames >= requiredStableFrames {
          stableDetections.append(StableDetection(
            tagId: tagId,
            poseMatrix: track.smoothedPose,
            centerX: result.centerX,
            centerY: result.centerY,
            hamming: Int(result.hamming),
            decisionMargin: result.decisionMargin,
            consecutiveFrames: track.consecutiveFrames,
            isNewlyStable: isNewlyStable
          ))
        }
      } else {
        // First time seeing this tag
        tagHistory[tagId] = TagTrack(
          consecutiveFrames: 1,
          smoothedPose: pose,
          lastCenter: (x: result.centerX, y: result.centerY)
        )
      }
    }

    // Reset tracking for tags not seen this frame
    for existingTagId in tagHistory.keys where !currentTagIds.contains(existingTagId) {
      tagHistory.removeValue(forKey: existingTagId)
    }

    return stableDetections
  }

  /// Exponential moving average of each matrix element.
  private func smoothPose(previous: [Double], current: [Double]) -> [Double] {
    guard previous.count == 16 && current.count == 16 else { return current }
    return zip(previous, current).map { prev, cur in
      prev * (1.0 - smoothingAlpha) + cur * smoothingAlpha
    }
  }

  /// Reset all tracking state (e.g. when session restarts).
  func reset() {
    tagHistory.removeAll()
  }
}

File: beachechoes/modules/beachechoes-ar/android/src/main/java/expo/modules/beachechoesar/AprilTagDetector.kt
package expo.modules.beachechoesar

/**
 * Wraps AprilTagJNI and adds multi-frame pose smoothing / stability gating.
 */
class AprilTagDetector {

    private val requiredStableFrames = 5
    private val smoothingAlpha = 0.3
    /** Default tag size in metres (6 inches ≈ 0.1524 m). */
    private val tagSizeMeters = 0.1524

    private var initialised = false
    private val tagHistory = mutableMapOf<Int, TagTrack>()

    data class TagTrack(
        var consecutiveFrames: Int = 0,
        var smoothedPose: FloatArray,
        var lastCenterX: Float = 0f,
        var lastCenterY: Float = 0f
    )

    data class StableDetection(
        val tagId: Int,
        val poseMatrix: List<Double>,
        val centerX: Double,
        val centerY: Double,
        val hamming: Int,
        val decisionMargin: Float,
        val consecutiveFrames: Int,
        val isNewlyStable: Boolean
    )

    fun init() {
        if (!initialised) {
            AprilTagJNI.nativeInit()
            initialised = true
        }
    }

    fun destroy() {
        if (initialised) {
            AprilTagJNI.nativeDestroy()
            initialised = false
        }
        tagHistory.clear()
    }

    /**
     * Run detection on a grayscale image.
     */
    fun detect(
        grayscale: ByteArray,
        width: Int,
        height: Int,
        fx: Double,
        fy: Double,
        cx: Double,
        cy: Double
    ): List<StableDetection> {
        if (!initialised) return emptyList()

        val rawResults = AprilTagJNI.nativeDetect(
            grayscale, width, height,
            fx, fy, cx, cy,
            tagSizeMeters
        )

        val currentTagIds = mutableSetOf<Int>()
        val stableDetections = mutableListOf<StableDetection>()

        for (row in rawResults) {
            val tagId = row[0].toInt()
            currentTagIds.add(tagId)

            val pose = row.sliceArray(5..20)

            val track = tagHistory[tagId]
            if (track != null) {
                track.consecutiveFrames++
                smoothPose(track.smoothedPose, pose)
                track.lastCenterX = row[1]
                track.lastCenterY = row[2]

                val isNewlyStable = track.consecutiveFrames == requiredStableFrames

                if (track.consecutiveFrames >= requiredStableFrames) {
                    stableDetections.add(StableDetection(
                        tagId = tagId,
                        poseMatrix = track.smoothedPose.map { it.toDouble() },
                        centerX = row[1].toDouble(),
                        centerY = row[2].toDouble(),
                        hamming = row[3].toInt(),
                        decisionMargin = row[4],
                        consecutiveFrames = track.consecutiveFrames,
                        isNewlyStable = isNewlyStable
                    ))
                }
            } else {
                tagHistory[tagId] = TagTrack(
                    consecutiveFrames = 1,
                    smoothedPose = pose.copyOf(),
                    lastCenterX = row[1],
                    lastCenterY = row[2]
                )
            }
        }

        // Remove tags not seen this frame
        tagHistory.keys.retainAll(currentTagIds)

        return stableDetections
    }

    fun reset() {
        tagHistory.clear()
    }

    private fun smoothPose(smoothed: FloatArray, current: FloatArray) {
        for (i in smoothed.indices) {
            smoothed[i] = (smoothed[i] * (1.0 - smoothingAlpha) + current[i] * smoothingAlpha).toFloat()
        }
    }
}
